docs/cadence_migration_guide/type-annotations-guide.mdx

```diff
@@ -0,0 +1,73 @@
+---
+title: Cadence Type Annotations 1.0 Migration Guide
+sidebar_position: 5
+sidebar_label: Cadence Type Annotations 1.0 Guide
+---
+
+# Type Annotations in Cadence 1.0
+
+In addition to updating your contracts in reaction to the Core Contract changes in Cadence 1.0, 
+certain language changes in Cadence 1.0 will also require changes to type annotations in your contracts, 
+in particular type annotations on resource and struct fields. 
+These type updates are required to accurately reflect the way that the Cadence 1.0 data migrations will change
+the types of these fields' values, and the Cadence 1.0 upgrade validator will enforce that these upgrades are accurate. 
+
+## Restricted Types
+
+In Cadence 1.0, support for restricted types was dropped, and replaced with intersection types. 
+As such, any existing restricted types must be replaced with a different type. 
+Restricted typed-values are migrated according to a specific set of rules, and type updates must also mirror this. 
+
+* `AnyStruct{I}` and `AnyResource{I}` should be migrated to just `{I}`, as these types have identical behavior
+* For any other type `T`, `T{I}` should be migrated to `T`, as this is the most specific possible type that can go here
+* For any type `T`, `T{}` should be migrated to just `T`, as there is no support for empty intersection types
+
+So, for example, a value of type `Vault{Receiver}` should be migrated to just a `Vault` value.
+
+## Reference Types
+
+Reference types from contracts written in Cadence v0.42 will need to be given entitlements in order to retain the same functionality 
+in Cadence 1.0. 
+The Cadence 1.0 data migration will automatically grant the appropriate entitlements to stored values, 
+but any reference types that appear in your contracts will need to be manually updated. 
+
+The update you will need to perform involves changing each reference type to have the appropriate entitlements necessary to perform 
+the same operations in Cadence 1.0 that it previously could in Cadence v0.42. 
+The Cadence 1.0 upgrade validator will enforce that these upgrades are accurate. 
+
+The first basic concept necessary to understand the upgrade is the "entitlements function"; 
```

Comment:
This whole section from 38-end is very academic. It uses a lot of terms and examples that 90% of developers are not going to understand. Would it be possible to rewrite it in more laymans terms. It doesn't need to be 100% accurate wording, just getting the point across without getting too academic

Reply:
Agreed that the language here is pretty dry; however, at the end of the day developers need to come out of this section having an accurate understanding of how the entitlements function works, since the contract upgrade validator is going to reject any contract upgrades that don't grant the correct set of entitlements to its reference types. So in a sense, the wording does need to be 100% accurate. 

When you were upgrading the core contracts did you have a way of thinking about which entitlements to give which references that might be more approachable to more developers?

Reply:
I disagree with the assumption that developers need to understand how the entitlements function works. Developers need to understand what they need to do to make sure their contracts upgrade properly.

When I was upgrading the core contracts, I just went through each contract, looked at each resource type, and thought about a few things:

1. If this is a resource type that typically has a public capability, which fields and functions that were previously `pub` need to now have an entitlement so that people can't access them by downcasting?
2. In my example transactions, how do I need to update the capability issuing so that the same functionality is exposed.
3. For any fields, function arguments, or stored values that are a reference or a capability that require access to one of the entitled functions, what `auth(entitlement)` do I need to include in the type specification so that that authorized function is still accessible?

I didn't think about entitlements functions or sets or anything academic like that. I was purely thinking practically, and I'm pretty sure I got to the correct types that way. I don't know if that helps here, but let me know.

Reply:
Hmm, that sounds like a lot of thought, especially given that there is really only one "correct" type upgrade for any given reference type, as computed by the migration. In theory it should be possible to just scan a contract and a type to see what entitlements it would need, rather than having to do any reasoning about contract functionality. This "automatic" nature is what I'm trying to capture by describing the upgrade this way.

That said, given how mechanical this is, in a non-blockchain world I would want to automate this upgrade by automatically upgrading peoples' types for them. I understand why we cannot do this, however it might be useful to have some kind of LSP integration that would suggest the correct upgrade for people, rather than asking them to do any manual thinking.

Reply:
The contract update checker will require that types are updated to exactly one option of many possibilities, so would it be possible to preface the full explanation with something similar to "you need to update to what the update checker tells you to update to" (otherwise the update is rejected anyway)

Reply:
thats reasonable